### PR TITLE
Build/modifier applying effect

### PR DESCRIPTION
**Pull request by Jordan-B1**

### CODE REVIEW

**Review by TheVengeur — APPROVED** [review]

File: src/modifier/mod.rs [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+/// MeleeWeaponModifier: The melee weapon modifier
+///
+/// RangedWeaponModifier: The ranged weapon modifier
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Modifier {
+    Global(GlobalModifier),
+    Offensive(OffensiveModifier),
+    Defensive(DefensiveModifier),
+    MeleeWeapon(MeleeWeaponModifier),
+    RangedWeapon(RangedWeaponModifier),
```
[the remark this answers is not in the document]
Comment (Pepiloto) [reply src/modifier/mod.rs:105]: If you still want to make a difference you have to create an interface (cpp like) to create a weapon "class" then create "subclasses" for ranged and melee weapons

File: src/fight/mod.rs [hunk 2]
```diff
@@ -39,15 +39,15 @@ fn compute_mean_case(
     attacking_regiment: &regiment::Regiment,
     defending_regiment: &regiment::Regiment,
 ) -> (usize, f64) {
-    let attacking_stats: &model::Stats = attacking_regiment.get_model().get_stats();
-    let defending_stats: &model::Stats = defending_regiment.get_model().get_stats();
+    let attacking_stats: &model::Stats = &attacking_regiment.get_model().get_stats();
+    let defending_stats: &model::Stats = &defending_regiment.get_model().get_stats();
```
Comment (Pepiloto) [inline src/fight/mod.rs:43]: Why this change ?

File: src/fight/mod.rs [hunk 3]
```diff
@@ -131,13 +131,13 @@ fn compute_case(
     defending_regiment: &regiment::Regiment,
     case: &ComputeCase,
 ) -> (usize, f64) {
-    let attacking_stats: &model::Stats = attacking_regiment.get_model().get_stats();
+    let attacking_stats: &model::Stats = &attacking_regiment.get_model().get_stats();
```
Comment (Pepiloto) [inline src/fight/mod.rs:134]: Why this change ?

File: src/fight/mod.rs [hunk 4]
```diff
@@ -131,13 +131,13 @@ fn compute_case(
...
     let nb_touch: usize =
         (attacking_stats.get_attack() as f64 * 1.5 * attacking_regiment.get_cols() as f64).round()
             as usize;
     let wound_probability: f64 = computation_tools::compute_damage_probability(
         attacking_stats,
-        defending_regiment.get_model().get_stats(),
+        &defending_regiment.get_model().get_stats(),
```
Comment (Pepiloto) [inline src/fight/mod.rs:140]: Why this change ?

File: src/fight/mod.rs [hunk 5]
```diff
@@ -217,8 +217,8 @@ fn create_prediction(
         ComputeCase::MEAN => ComputeCase::MEAN,
     };
     let fastest: u8 = computation_tools::find_the_fastest(
-        attacking_regiment.get_model().get_stats(),
-        defending_regiment.get_model().get_stats(),
+        &attacking_regiment.get_model().get_stats(),
+        &defending_regiment.get_model().get_stats(),
```
Comment (Pepiloto) [inline src/fight/mod.rs:221]: Why this change ?

File: src/math_tools/mod.rs [hunk 6]
```diff
@@ -132,4 +149,14 @@ mod tests {
...
+
+    #[test]
+    fn test_signed_adder() {
+        let res: usize = super::safe_add_signed_unsigned(5, 2);
+        assert_eq!(res, 7);
+        let res: usize = super::safe_add_signed_unsigned(5, -2);
+        assert_eq!(res, 3);
+        let res: usize = super::safe_add_signed_unsigned(5, -10);
+        assert_eq!(res, 0);
+    }
```
Comment (Pepiloto) [inline src/math_tools/mod.rs:161]: What happen for -5 and 0 ?

File: src/model/mod.rs [hunk 7]
```diff
@@ -249,41 +251,106 @@ impl Stats {
...
-    requirements: Vec<String>,
-}
+    /// Buff the Stats with a GlobalModifier
+    ///
+    /// ## Parameters
+    /// &modifier (GlobalModifier): The GlobalModifier to buff the Stats with
+    ///
+    /// ## Returns
+    /// &Stats: The Stats buffed
+    fn buff_global_modifier(&mut self, modifier: &modifier::GlobalModifier) -> &Stats {
```
Comment (Pepiloto) [inline src/model/mod.rs:262]: This function name is misleading I think people will believe it will buf the global modifier but instead it buff the stats.
And why do we need the return value ? Just change the Stats maybe ?

File: src/model/mod.rs [hunk 8]
```diff
@@ -249,41 +251,106 @@ impl Stats {
...
+    }
+
+    /// Buff the Stats with a DefensiveModifier
+    ///
+    /// ## Parameters
+    /// &modifier (DefensiveModifier): The DefensiveModifier to buff the Stats with
+    ///
+    /// ## Returns
+    /// &Stats: The Stats buffed
+    fn buff_defensive_modifier(&mut self, modifier: &modifier::DefensiveModifier) -> &Stats {
```
Comment (Pepiloto) [inline src/model/mod.rs:298]: Same as above

File: src/model/mod.rs [hunk 9]
```diff
@@ -249,41 +251,106 @@ impl Stats {
...
+    }
+
+    /// Buff the Stats with the given melee modifier
+    ///
+    /// ## Parameters
+    /// (&modifier::MeleeWeaponModifier) modifier : The modifier to apply
+    ///
+    /// ## Returns
+    /// &Stats: The Stats buffed
+    fn buff_melee_weapon(&mut self, weapon: &modifier::MeleeWeaponModifier) -> &Stats {
```
Comment (Pepiloto) [inline src/model/mod.rs:315]: Same as above

File: src/model/mod.rs [hunk 10]
```diff
@@ -249,41 +251,106 @@ impl Stats {
...
+    }
+
+    /// Buff the Stats with the given ranged modifier
+    ///
+    /// ## Parameters
+    /// (&modifier::RangedWeaponModifier) modifier : The modifier to apply
+    ///
+    /// ## Returns
+    /// &Stats: The Stats of the Model with the modifier applied
+    fn buff_ranged_weapon(&mut self, weapon: &modifier::RangedWeaponModifier) -> &Stats {
```
Comment (Pepiloto) [inline src/model/mod.rs:331]: Same as above

**Review by Pepiloto — COMMENTED** [review]
Did you try the `add_new_unit.py` script to be sure it still worked ?

**Review by H4ss — APPROVED** [review]

**Comment by Jordan-B1** [comment]
> Did you try the `add_new_unit.py` script to be sure it still worked ?

It was not working but I fixed it :+1:

File: src/fight/mod.rs (hunk 2 above)
Comment (Jordan-B1) [reply src/fight/mod.rs]: In the function I was creating a copy of the stats struct but finally changed my way of handling the modifier so this change is no longer existing

File: src/math_tools/mod.rs (hunk 6 above)
Comment (Jordan-B1) [reply src/math_tools/mod.rs:161]: The first parameter is a usize, it can't be -5, the code won't compile

File: src/model/mod.rs (hunk 7 above)
Comment (Jordan-B1) [reply src/model/mod.rs]: Changed it :heavy_check_mark:

File: src/modifier/modifiers_id.rs [hunk 11]
```diff
@@ -0,0 +1,3 @@
+pub const _RANGEDSTATID: &str = "a00c-d586-ee68-ed21";
+
+pub const _MELEESTATID: &str = "a32f-208a-be3d-ad8d";
```
[the remark this answers is not in the document]
Comment (Jordan-B1) [reply src/modifier/modifiers_id.rs]: Oh a file I forgot to delete, it's no longer here

**Review by TheVengeur — APPROVED** [review]

File: src/model/mod.rs [hunk 12]
```diff
@@ -249,41 +251,73 @@ impl Stats {
...
-pub struct Modifier {
-    stat: Stats,
-    bonus: bool,
-    nb_dice: usize,
-    requirements: Vec<String>,
-}
+    /// Buff the Stats with a GlobalModifier
+    ///
+    /// ## Parameters
+    /// &modifier (GlobalModifier): The GlobalModifier to buff the Stats with
```
Comment (Pepiloto) [inline src/model/mod.rs:258]: It not `&modifier` but `(&GlobalModifier)`

File: src/model/mod.rs [hunk 13]
```diff
@@ -249,41 +251,73 @@ impl Stats {
...
+        self.advance = math_tools::safe_add_signed_unsigned(self.advance, modifier.get_advance());
+        self.march = math_tools::safe_add_signed_unsigned(self.march, modifier.get_march());
+        self.discipline =
+            math_tools::safe_add_signed_unsigned(self.discipline, modifier.get_discipline());
+    }
+
+    /// Buff the Stats with a OffensiveModifier
+    ///
+    /// ## Parameters
+    /// &modifier (OffensiveModifier): The OffensiveModifier to buff the Stats with
```
Comment (Pepiloto) [inline src/model/mod.rs:269]: It not &modifier but (&OffensiveModifier)

File: src/model/mod.rs [hunk 14]
```diff
@@ -249,41 +251,73 @@ impl Stats {
...
-impl Modifier {
-    /// Create a new Modifier using Stats, number of dice and the requirements to apply the Modifier
+    /// Buff the Stats with a DefensiveModifier
     ///
     /// ## Parameters
-    /// (Stats) stats: The stats affected by the modifier
-    /// (bool) bonus: A boolean value that indicate if the effect of the modifier should be added or substract from the model who's concerned
-    /// (usize) nb_dice: The number of supplement rolls granted by the modifier
-    /// (Vec<String>) requirements: The requirements for the Modifier to apply (as an array of flags)
+    /// &modifier (DefensiveModifier): The DefensiveModifier to buff the Stats with
```
Comment (Pepiloto) [inline src/model/mod.rs:286]: It not &modifier but (&DefensiveModifier)

File: src/model/mod.rs [hunk 15]
```diff
@@ -249,41 +251,73 @@ impl Stats {
...
+        self.defense = math_tools::safe_add_signed_unsigned(self.defense, modifier.get_defense());
+        self.resilience =
+            math_tools::safe_add_signed_unsigned(self.resilience, modifier.get_resilience());
+        self.armour = math_tools::safe_add_signed_unsigned(self.armour, modifier.get_armour());
+    }
+
+    /// Buff the Stats with the given weapon modifier
+    ///
+    /// ## Parameters
+    /// (&modifier::WeaponModifier) modifier : The modifier to apply
```
Comment (Pepiloto) [inline src/model/mod.rs:299]: Why change way to make documentation for this function ?

File: src/model/mod.rs [hunk 16]
```diff
@@ -249,41 +251,73 @@ impl Stats {
...
+            self.armour_penetration,
+            weapon.get_armour_penetration(),
+        );
+        self.strength = math_tools::safe_add_signed_unsigned(self.strength, weapon.get_strength());
+    }
+
+    /// Apply the given modifier to the Stats
+    ///
+    /// ## Parameters
+    /// &modifier (Modifier): The Modifier to apply
```
Comment (Pepiloto) [inline src/model/mod.rs:311]: It not &modifier but (&Modifier)

File: src/model/mod.rs [hunk 17]
```diff
@@ -249,41 +251,73 @@ impl Stats {
...
     ///
     /// ## Return
-    /// Modifier: The Modifier created
-    pub fn new(stat: Stats, bonus: bool, nb_dice: usize, requirements: Vec<String>) -> Modifier {
-        Modifier {
-            stat,
-            bonus,
-            nb_dice,
-            requirements,
+    /// &Stats: The Stats buffed
```
Comment (Pepiloto) [inline src/model/mod.rs:314]: There is no return in this function I think

File: src/model/mod.rs [hunk 18]
```diff
@@ -309,20 +346,39 @@ impl Model {
...
+            modifiers,
+        }
     }
-    /// Get the Stats of the Model
+    /// Get the Stats of the Model taking account of the modifiers
     ///
     /// ## Return
-    /// Stats: The Stats of the Model
+    /// &Stats: The Stats of the Model
     pub fn get_stats(&self) -> &Stats {
```
Comment (Pepiloto) [inline src/model/mod.rs:364]: The name of the function isn't right it doesn't return the stats but the `boosted_stats`

File: src/model/mod.rs [hunk 19]
```diff
@@ -309,20 +346,39 @@ impl Model {
...
+        &self.boosted_stats
+    }
+
+    /// Get the Stats of the Model without taking account of the modifiers
+    ///
+    /// ## Return
+    /// &Stats: The Stats of the Model
+    pub fn get_pure_stats(&self) -> &Stats {
```
Comment (Pepiloto) [inline src/model/mod.rs:372]: `pure` isn't really intuitive maybe just `get_stats` is clearer

File: src/regiment/mod.rs [hunk 20]
```diff
@@ -40,8 +40,8 @@ impl Regiment {
     ///
     /// (usize) nb_models: The number of models in the regiment
     ///
-    /// Option<usize> regiment_health_point: The number of health points that left to the regiment.
-    /// If None, the default value will be (the number of model * the health_point stats of the model)
+    /// Option<usize> regiment_health_points: The number of health points that left to the regiment.
+    /// If None, the default value will be (the number of model * the health_points stats of the model)
```
Comment (Pepiloto) [inline src/regiment/mod.rs:44]: Why add Option here ?
The regiment health point can't be just `nb_rows * nb_cols * model_health_points` or `nb_model * model_health_points` and reduce during the game ?

**Review by Pepiloto — CHANGES_REQUESTED** [review]

File: src/regiment/mod.rs (hunk 20 above)
Comment (Jordan-B1) [reply src/regiment/mod.rs:44]: we talked about it during a meeting. As there are some models that have more than one HP.
without this regiment_health_point  we can't take account when a unit lost 1 HP but is not dead. The code is already implemented and was validated in a previous PR :|

**Review by Pepiloto — APPROVED** [review]

**Review by H4ss — DISMISSED** [review]
